Initial inference graph API implementation

yuzisun opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #846 

Please see the detailed proposal in https://docs.google.com/document/d/13VHfOxa72pgoy5Eg5c-gGHZfEL7uBDc7rTLa2pAi4Ko

**Special notes for your reviewer**:

1. tested examples

```yaml
---
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "sklearn-iris"
spec:
  predictor:
    sklearn:
      storageUri: "gs://kfserving-examples/models/sklearn/1.0/model"
---
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "xgboost-iris"
spec:
  predictor:
    xgboost:
      storageUri: "gs://kfserving-examples/models/xgboost/iris"
---
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: ensemble-model
spec:
  nodes:
    ensembleModel:
      routerType: Ensemble
      routes:
        - service: sklearn-iris
        - service: xgboost-iris
---
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: model-chainer
spec:
  nodes:
    root:
      routerType: Single 
      routes:
        - service: sklearn-iris
      nextRoutes:
        - nodeName: treeModel
          data: $request
    treeModel:
      routerType: Single
      routes:
        - service: xgboost-iris
---
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: splitter-model
spec:
  nodes:
    TrafficSplit:
      routerType: Splitter 
      routes:
        - service: sklearn-iris
          weight: 20
        - service: xgboost-iris
          weight: 80
```

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
alpha version of inference graph with basic model chaining, traffic splitting, ensemble capability.
```

yuzisun committed:
Add inference graph API

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Implement inference graph API

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

pugangxa commented on pkg/controller/v1alpha1/inferencegraph/controller.go:
```diff
@@ -0,0 +1,266 @@
+		}
+		return reconcile.Result{}, err
+	}
+
```
Question:
Is there any validation on the graph configuration? I mean should the service url being existing inferenceservice endpoint or just any string is accepted? 
Also I saw in the router image there's no validation too so it's just behave like proxy and user need make the request following our request spec? 
In fact if there's no any validation user may chain any knative service and make calls as long as the request is valid for the service accordingly. That may be an additional flexibility.

yuzisun replied:
The idea is to support both `InferenceService` name and url, if `InferenceService` name is provided then the controller fills in the url from the status object. Will also add the graph validation and graph status based on the individual inference service status.

Graph orchestrator serves as the proxy and  I think the request validation should be done in the destination service and the proxy can propagate the error back.

pugangxa replied:
That's clear to me now, thanks for response and look forward to the feature.

yuzisun replied:
if `InferenceService` name is provided we resolve to the service url and in the router it is always use service url to invoke the request.

pugangxa commented on pkg/controller/v1alpha1/inferencegraph/controller.go:
```diff
@@ -0,0 +1,266 @@
+		return reconcile.Result{}, err
+	}
+	// Create service if does not exist
+	desired := createKnativeService(graph.ObjectMeta, graph, routerConfig)
```
Before we already dropped the hard requirement for knative by deployment and HPA, so for graph do we consider add adaptation to this? Or I think this will add back the dependency on knative.

yuzisun replied:
We still support both Knative and raw deployment mode, will add the raw deployment handling as well.

pugangxa replied:
Yeah, great!

Iamlovingit commented on config/rbac/role.yaml:
```diff
@@ -264,3 +264,23 @@ rules:
   - patch
   - update
+- apiGroups:
+  - serving.kubeflow.org
```
`apiGroups` should be `serving.kserve.io`

Commit:
add validation and splitter

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

rachitchauhan43 commented:
@yuzisun : When is this PR scheduled to get merged ? Will it be part of v0.9 release ?

yuzisun committed:
Fix inference graph status

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Refactor knative reconciler

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Resolve service urls

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun commented on cmd/router/main.go:
```diff
@@ -0,0 +1,165 @@
+		var res map[string]interface{}
+		json.Unmarshal(responseBytes, &res)
+		response = res
+	} else if currentStep.RouterType == v1alpha1.Switch {
```
This has not been implemented yet, we plan to use [k8s.io/client-go/util/jsonpath](http://k8s.io/client-go/util/jsonpath) to evaluate the jsonpath and route the request based on condition

njhill replied:
I remember someone saying that we might use some other json expression library for this since k8s jsonpath is quite limited?

yuzisun replied:
we are going to use gjson https://github.com/tidwall/gjson which is used by argo for condition check.

yuzisun committed:
Improve logging

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun commented:
> @yuzisun : When is this PR scheduled to get merged ? Will it be part of v0.9 release ?

@rachitchauhan43 this PR is ready for review, please help test out if you are interested, the plan is to release the alpha version in 0.9.

rachitchauhan43 commented:
> > @yuzisun : When is this PR scheduled to get merged ? Will it be part of v0.9 release ?
> 
> @rachitchauhan43 this PR is ready for review, please help test out if you are interested, the plan is to release the alpha version in 0.9.

@yuzisun : Happy to help out in testing this. Where can I start at to test it out ? Or any particular scenarios documented to test it out ?

yuzisun commented:
@njhill @pvaneck Would like to get your feedback on this.

Commit:
add switch and root node configuration.

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update docs/samples/graph/README.md

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

pvaneck commented on Makefile:
```diff
@@ -183,9 +193,15 @@ docker-push:
 	docker build -f agent.Dockerfile . -t ${KO_DOCKER_REPO}/${AGENT_IMG}
 
+docker-build-router:
+	docker build -f router.Dockerfile . -t ${KO_DOCKER_REPO}/${ROUTER_IMG}
```
Doesn't look like `router.Dockerfile` was added?

pvaneck commented on cmd/manager/main.go:
```diff
@@ -158,6 +159,20 @@ func main() {
+		Client: mgr.GetClient(),
+		Log:    ctrl.Log.WithName("v1alpha1Controllers").WithName("InferenceGraph"),
+		Scheme: mgr.GetScheme(),
+		Recorder: eventBroadcaster.NewRecorder(mgr.GetScheme(), v1.EventSource{Component: "InferenceGraph	Controller"}),
```
Should the EventSource Component be `v1alpha1Controllers`?

pvaneck commented on docs/samples/graph/README.md:
```diff
@@ -0,0 +1,251 @@
+## **2. KServe Inference Graph** 
+![image](graph.png)
+### **2.1 Inference Graph**
+As above image shows, inference graph is made up of a list of `nodes` , and each `node` consists of several `isvcs`. Every graph must have a root node named `root`. When an inference request hits the graph it excutes the `root` node from the DFS, if the graph has other `nodes`, it will pass the `$request` or `$response` of the root service as input data to the `next node`. There are four types of ...
```
```suggestion
As the above image shows, an inference graph is made up of a list of `nodes`, and each `node` consists of several `isvcs`. Every graph must have a root node named `root`. When an inference request hits the graph, it executes the `root` node from the DAG. If the graph has other `nodes`, it will pass the `$request` or `$response` of the root service as input data to the `next node`. There are four `node` types that are supported: ***Single***, ***Switch***, ***Ensemble***, ***Splitter***.
```

pvaneck commented on docs/samples/graph/README.md:
```diff
@@ -0,0 +1,251 @@
+
+
+### **2.2 Single Node**
+**Single Node** makes user can connect 2 isvcs in a sequence relationship. The
```
```suggestion
**Single Node** allows users to connect two `isvcs` in a sequence relationship. The
```

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+...
+```
+### **2.3 Switch Node**
+**Switch Node** makes user can select an `isvc` to handle the request by setting the `condition`. Usually, user doesn't need to set the `serviceURL`, kserve will fill it with `isvc.status.address.URL` of `isvc`, but if you want to specified the `serviceURL`, you can set it manually. 
````
```suggestion
**Switch Node** allows users to select an `isvc` to handle the request by setting the `condition`. Usually, users don't need to set the `serviceURL` as KServe will fill it in with `isvc.status.address.URL` of the `isvc`, but if desired, `serviceURL` can be specified explicitly.
```

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+  - service: xgboost-iris
+```
+### **2.5 Splitter Node**
+**Splitter Node** make user can  
````
Looks like a description here was mistakenly left out? I'm assuming this would say something like:

**Splitter Node** allows users to split traffic to multiple `isvcs` using a weighted distribution

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+NAME            URL                                                  READY   AGE
+model-chainer   http://model-chainer.default.10.166.15.29.sslip.io   True    5s
+```
+3. Tesing `graph`.
````
s/Tesing/Testing/

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+NAME            URL                                                  READY   AGE
+model-switch     http://model-switch.default.10.166.15.29.sslip.io     True    35s
+```
+3. Tesing `graph`.
````
s/Tesing/Testing/

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+```
+3. Tesing `graph`.
+```shell
+curl http://model-switch.default.10.166.15.29.sslip.io
````
I think an example where the user passes in an explicit target (i.e. target == "blue") would be more beneficial in terms of showcasing usage.

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+NAME            URL                                                  READY   AGE
+splitter-model   http://splitter-model.default.10.166.15.29.sslip.io   True    15m
+```
+3. Tesing `graph`.
````
s/Tesing/Testing/

pvaneck commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+NAME            URL                                                  READY   AGE
+ensemble-model   http://ensemble-model.default.10.166.15.29.sslip.io   True    15m
+```
+3. Tesing `graph`.
````
s/Tesing/Testing/

pvaneck commented on docs/samples/graph/sequence.yaml:
```diff
@@ -0,0 +1,37 @@
+        - service: sklearn-iris
+      nextRoutes:
+        - nodeName: treeModel
+          data: $request
```
Would be cool to also have a sample where root.nextRoutes.data is $response instead of $request (doesn't need to be this PR)

njhill replied:
Related to my other comment in the code corresponding to this, what will happen to the output (response) from the `sklearn-iris` service here? It seems like it will essentially be dropped, in which case why even include it?

yuzisun replied:
this example is over-simplified, often the time you would need to add condition based on the response and then forward the request to the next model.

pvaneck reviewed (commented):
Hey, @yuzisun. This looks great so far. Did an initial review with some comments inline. 

Also noticed that there are several line changes to the servingruntime and isvc crd yamls. I'm assuming because of differing versions of yq shifting the indentation when `make manifests` is run, creating quite huge diffs. In the storage spec PR a while back, an older version of yq was used in the generation, but now, it looks like a newer version of 'yq' is used, re-shifting the indentation.

Should we have the crd regeneration done in a separate PR to keep this PR cleaner? And should we do something like we do with ModelMesh-Serving where we [perform](https://github.com/kserve/modelmesh-serving/blob/main/scripts/develop.sh) tests/generation using a [developer image](https://github.com/kserve/modelmesh-serving/blob/main/Dockerfile.develop) with correct dependencies installed?

rachitchauhan43 commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+``` 
+***Expect result***
+```shell
+{"http://sklearn-iris.default.svc.cluster.local/v2/models/sklearn-iris/infer":{"predictions":[1,1]},"http://xgboost-iris.default.svc.cluster.local/v2/models/xgboost-iris/infer":{"predictions":[1,1]}}
````
Why keys for predictions are complete service URLs ? Why not just use the service name ike `sklearn-iris` and `xgboost-iris` as the key ?
Won't output be more readable like that ? 
```
{
  "sklearn-iris": {
    "predictions": [
      1,
      1
    ]
  },
  "xgboost-iris": {
    "predictions": [
      1,
      1
    ]
  }
}
```

rachitchauhan43 replied:
Think about the scenario when ensemble’s output is passed onto downstream model in the graph. Then working on the json with URL as keys is not going to a great developer/user experience.

yuzisun replied:
though service name is optional and in case only service url is provided we won't be able to populate that, but agreed we can at least optimize for the service name case.

yuzisun replied:
I have added a new field `stepName` for this, so the merged response will be populated with `stepName` as the key instead of using `serviceUrl`.

rachitchauhan43 replied:
`stepName` field would work fine too.

rachitchauhan43 commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+var log = logf.Log.WithName("InferenceGraphRouter")
+
+func callService(serviceUrl string, input []byte, res chan<- []byte) error {
+	resp, err := http.Post(serviceUrl, "application/json", bytes.NewBuffer(input))
```
I created an IG with a non-ISVC node and IG was able to invoke that.
```
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: ext-service-caller
spec:
  nodes:
    root:
      routerType: Single
      routes:
        - serviceUrl: http://helloworld.kserve-test.svc.cluster.local:5000/hello
```

So, one thing we observed while invoking external services as a node in IG is that, IG will only make calls assuming those nodes are ISVCs hence the HTTP verb is always POST.(https://github.com/yuzisun/kserve/blob/graph/cmd/router/main.go#L42)

But think about scenarios where, someone would want to invoke some external services like AWS or Google cloud services or in-house feature store service they have in their org that has other semantics like HTTP GET to fetch data.

In that case what would be the right solution ? One solution I see is, creating a pseudo-isvc that internally just makes calls to other services.

Can we think of something more general ? Or allowing extending how the node should be invoked ?

rachitchauhan43 replied:
As per our discussion in community meting, transformers should be used for this prupose.

njhill commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+		jobs[routeTo.NodeName] = job
+		if router, ok := graph.Nodes[routeTo.NodeName]; ok {
+			if routeTo.Data == "$request" {
+				go routeStep(routeTo.NodeName, router, graph, input, job)
```
This case seems odd, basically we must be throwing away the result of the prior step (`jsonRes`) so why perform it at all? (unless it has some other side effect like payload logging, maybe that's reasonable)

yuzisun replied:
there could be case that you want to add the condition based on the response but still send the input to next model with the previous request, for example you first classify the text with high level category and then send the original text as input to next model to classify sub categories based on the previous model prediction result.

njhill commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+	}
+	// process outgoing edges
+	jobs := map[string]chan []byte{}
+	for _, routeTo := range currentStep.NextRoutes {
```
The multiple nextRoutes logic here doesn't appear to be documented. The docs/examples in this PR imply there should only be a single one.

Reading what is implemented here, it looks similar/equivalent to the ensemble case, except that the parallel branches are other nodes rather than services, and there's no way to do any kind of post-processing of the result (I guess it would have to be the output of the inference graph as a whole).

yuzisun replied:
right, this is for sub-pipeline ensemble case, e.g parallel branches for different model pipelines. I liked your idea in the [discussion](https://github.com/kserve/kserve/discussions/2170), it is more general and you can also apply the condition and weights for nodes.
```
     | - transformer1 -> model1 -|
-->                              | ->
     | - transformer2 -> model2 -|
```

njhill commented on docs/samples/graph/switch.yaml:
```diff
@@ -0,0 +1,51 @@
+      routerType: Switch
+      routes:
+      - service: blue
+        serviceUrl: http://blue.default.10.166.15.29.sslip.io/splitter
```
The CRD doc states `InferenceService URL, mutually exclusive with Service`, so this example appears to be invalid?

yuzisun replied:
right, should remove the serviceUrl

njhill commented on docs/samples/graph/README.md:
````diff
@@ -0,0 +1,251 @@
+```shell 
+kubectl apply -f splitter.yaml
+```
+2. Waiting for `isvc` and `graph` up.
````
```suggestion
2. Waiting for `isvc` and `graph` to come up.
```

njhill commented on pkg/controller/v1alpha1/inferencegraph/controller.go:
```diff
@@ -0,0 +1,224 @@
+
+	r.Log.Info("Reconciling inference graph", "apiVersion", graph.APIVersion, "graph", graph.Name)
+	configMap := &v1.ConfigMap{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: constants.InferenceServiceConfigMapName, Namespace: constants.KServeNamespace}, configMap)
```
Can use the passed-in `ctx` for this (and others below).

alexagriffith replied:
also question: why not use something like 
```
ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
defer cancel()
```

for the client call?

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,252 @@
+//       - service: cat-dog-classifier
+//       nextRoutes:
+//         nodeName: breed-classifier
+//         data: "{ $input }"
```
Syntax for this looks inconsistent with other code/docs. (`{ $input }` vs `$request`)

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,252 @@
+//   name: canary-route
+// spec:
+//   nodes:
+//     mymodel:
```
There's no node named `root` in these examples

njhill commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+	start, end := 0, 0
+
+	for _, route := range routes {
+		start += end
```
I think this line is meant to be `start = end`?

Also there is no need to keep track of `start` at all, can just remove that var and related check.

njhill reviewed (commented):
@yuzisun thanks for this and sorry for taking so long to look at it closely, I've taken a (non-exhaustive) first pass and added some comments.

The design doc appears to be a bit out of date, there's inconsistencies between the design doc, examples/doc in the PR, and actual implemented code. The resulting API seems a bit muddled to me, and I think has some limitations in terms of the kind of compositions possible.

I have some thoughts about how the API design could be streamlined a bit, I've created a discussion for those [here](https://github.com/kserve/kserve/discussions/2170) to avoid derailing this PR.

shrinandj commented:
I'd love to try this out. Any idea when this could get merged?

alexagriffith commented on config/webhook/manifests.yaml:
```diff
@@ -108,3 +108,30 @@ webhooks:
+          - CREATE
+          - UPDATE
+        resources:
+          - inferencegraphs
```
nit: ied complains no line at eof

alexagriffith replied:
this happens in a couple of files

alexagriffith commented on pkg/apis/serving/v1alpha1/inference_graph_validation.go:
```diff
@@ -0,0 +1,118 @@
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (ig *InferenceGraph) ValidateUpdate(old runtime.Object) error {
+	validatorLogger.Info("validate update", "name:", ig.Name)
```
nit: can we remove `:` in `name:` to be consistent with the other logging

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+		return nil
+	}
+	for _, route := range routes {
+		//new jsonpath object
```
obvious comment - could delete it

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+	for _, route := range routes {
+		//new jsonpath object
+		j := jsonpath.New("Parser")
+		//j.AllowMissingKeys(true)
```
did you mean to leave this here?

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+	//remove whitespaces
+	str := strings.Replace(origin, " ", "", -1)
+	//remove {}
+	str = str[1 : len(str)-1]
```
do we want to validate that the leading and trailing brackets exist?

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+			json.Unmarshal(responseBytes, &res)
+			response[name] = res
+		}
+	} else { //routeType == Single
```
just to be explicit, could we change else to `else if currentStep.RouterType == v1alpha1.Single` ? so that when the code changes, it is clear and if something is changed, it will break if it doesn't match one of the conditions.

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+	log.Info("current step", "nodeName", nodeName, "URL", currentStep.Routes[0].ServiceUrl)
+	defer timeTrack(time.Now(), nodeName)
+	response := map[string]interface{}{}
+	//For splitter and ABNTest call virtual service
```
nit: can you add a bit of "why" to this comment? like - why is splitter different here?

alexagriffith commented on docs/samples/graph/README.md:
```diff
@@ -0,0 +1,251 @@
+# **Inference Graph**
+## **1. Problem** 
+
+Often the time on a production inference pipeline, it is not a single inference service, models need to be chained together to produce the final prediction result. For example a face recognition pipeline may need to find the face area first and then compute the features of the faces to match the face database, these two models are depending on each other and the first model’s output is the second ...
```
```suggestion
Production inference pipelines are often composed of multiple inference services, and models need to be chained together to produce the final prediction result. For example, a face recognition pipeline may need to find the face area first and then compute the features of the faces to match the face database. These two models depend on each other, and the first model’s output is the second model’s input. Another example is the NLP pipeline. It is very common to do some document classification and then have downstream tasks for named entity detection or text summary.[[...]](https://docs.google.com/document/d/13VHfOxa72pgoy5Eg5c-gGHZfEL7uBDc7rTLa2pAi4Ko/edit#heading=h.x9snb54sjlu9)
```

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+
+func pickupRoute(routes []v1alpha1.InferenceRoute) *v1alpha1.InferenceRoute {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	//generate num [0,100)
```
nit: can you add "why" for this comment too? why do we need to gen random num? I think that will add value to the code/comment

alexagriffith commented on cmd/router/main.go:
```diff
@@ -0,0 +1,248 @@
+		return nil, err
+	}
+	var data Input
+	data.Items = append(data.Items, inputData)
```
if we set data here to the zero val of Input, why do we need to append instead of say 

```
data.Items := inputData
```
?

yuzisun committed:
Add inference router dockerfile

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Fix comments and error handling

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun commented:
> Hey, @yuzisun. This looks great so far. Did an initial review with some comments inline.
> 
> Also noticed that there are several line changes to the servingruntime and isvc crd yamls. I'm assuming because of differing versions of yq shifting the indentation when `make manifests` is run, creating quite huge diffs. In the storage spec PR a while back, an older version of yq was used in the generation, but now, it looks like a newer version of 'yq' is used, re-shifting the indentation.
> 
> Should we have the crd regeneration done in a separate PR to keep this PR cleaner? And should we do something like we do with ModelMesh-Serving where we [perform](https://github.com/kserve/modelmesh-serving/blob/main/scripts/develop.sh) tests/generation using a [developer image](https://github.com/kserve/modelmesh-serving/blob/main/Dockerfile.develop) with correct dependencies installed?

Sounds good, we can also do something like https://github.com/kserve/kserve/blob/master/hack/verify-codegen.sh to verify if the PR generates the files correctly by comparing the generation in CI.

yuzisun committed:
Updating single routerType to sequence for more flexisble composition

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

rachitchauhan43 commented:
Did observe this bug while testing:

Created this:
```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: custom-model
spec:
  predictor:
    containers:
      - name: kserve-container
        image: rachitchauhan885/custom-model:v4
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
---
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: hist-feature-fetcher
spec:
  predictor:
    containers:
      - name: kserve-container
        image: rachitchauhan885/hist-feature-fetcher:v1
        args:
          - --model_name
          - hist-feature-fetcher
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
---
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: latest-feature-fetcher
spec:
  predictor:
    containers:
      - name: kserve-container
        image: rachitchauhan885/latest-feature-fetcher:v1
        args:
          - --model_name
          - latest-feature-fetcher
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
---
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: qbo-in
  namespace: rachit
spec:
  predictor:
    containers:
      - name: kserve-container
        image: rachitchauhan885/qbo-in:v1
        args:
          - --model_name
          - qbo-in
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
---
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: qbo-us
  namespace: rachit
spec:
  predictor:
    containers:
      - name: kserve-container
        image: rachitchauhan885/qbo-us:v1
        args:
          - --model_name
          - qbo-us
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
---
apiVersion: serving.kserve.io/v1alpha1
kind: InferenceGraph
metadata:
  name: all-in-one
spec:
  nodes:
    root:
      routerType: Switch
      routes:
      - service: qbo-in
        condition: "{.input.country_code == \"IN\"}"
      - service: qbo-us
        condition: "{.input.country_code == \"US\"}"
      nextRoutes:
      - nodeName: feature-fetcher
        data: $response
    feature-fetcher:
      routerType: Ensemble
      routes:
      - service: hist-feature-fetcher
      - service: latest-feature-fetcher
      nextRoutes:
      - nodeName: txn-categorizer
        data: $response
    txn-categorizer:
      routerType: Single
      routes:
      - service: custom-model
```

The final output on calling all-in-one-graph is:
```
{
  "feature-fetcher": {
    "txn-categorizer": {
      "predictions": {
        "http://hist-feature-fetcher.rachit.svc.cluster.local/v1/models/hist-feature-fetcher:predict": {
          "historical_features": {
            "last_tax_year_status": "Single",
            "predictions": "From IN model"
          }
        },
        "http://latest-feature-fetcher.rachit.svc.cluster.local/v1/models/latest-feature-fetcher:predict": {
          "latest_features": {
            "predictions": "From IN model",
            "this_tax_year_status": "Married-Joint"
          }
        }
      }
    }
  }
}
```

But it should be this:
```{
  "txn-categorizer": {
    "predictions": {
      "feature-fetcher": {
        "http://hist-feature-fetcher.rachit.svc.cluster.local/v1/models/hist-feature-fetcher:predict": {
          "historical_features": {
            "last_tax_year_status": "Single",
            "predictions": "From IN model"
          }
        },
        "http://latest-feature-fetcher.rachit.svc.cluster.local/v1/models/latest-feature-fetcher:predict": {
          "latest_features": {
            "predictions": "From IN model",
            "this_tax_year_status": "Married-Joint"
          }
        }
      }
    }
  }
}
```
The outer most keys are not correctly placed. Can you verify ?

njhill committed:
Some InferenceGraph code streamlining

- Avoid unnecessary goroutine use
- Avoid most unnecessary unmarshal+remarshals
- Deduplicate logic between router types
- Handle async errors in ensemble case
- Don't require stepName to be set
- Return http error response when applicable
- Simplify pickupRoute func
- Fix incorrect log parameter syntax
- Add webhook validation for unique step names and inference targets

Signed-off-by: Nick Hill <nickhill@us.ibm.com>
Signed-off-by: Dan Sun <dsun20@bloomberg.net>

njhill commented:
@yuzisun thanks for the latest changes, I really like the node/service unification and sequence addition. While reviewing them I did a bit of streamlining of the graph execution logic and added a bit more validation, see https://github.com/yuzisun/kserve/pull/8. I do have some further thoughts/comments, will add them soon.

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,253 @@
+
+	// Routes defines destinations for the current router node
+	// +optional
+	Routes []InferenceStep `json:"routes,omitempty"`
```
Should we rename this now?
```suggestion
	Steps []InferenceStep `json:"steps,omitempty"`
```

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,253 @@
+	NodeName string `json:"nodeName,omitempty"`
+
+	// named reference for InferenceService
+	Service string `json:"service,omitempty"`
```
WDYT about renaming this to `ServiceName` for clarity/consistency?

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,253 @@
+
+	// InferenceService URL, mutually exclusive with Service
+	// +optional
+	ServiceUrl string `json:"serviceUrl,omitempty"`
```
per go style convention
```suggestion
	ServiceURL string `json:"serviceUrl,omitempty"`
```

njhill commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,253 @@
+
+// +k8s:openapi-gen=true
+
+type InferenceTarget struct {
```
```suggestion
// Exactly one InferenceTarget field must be specified
type InferenceTarget struct {
```

njhill reviewed (commented):
A few renaming suggestions

yuzisun committed:
Fix inference graph spec

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Add condition tests

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update graph doc

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

pvaneck commented on pkg/apis/serving/v1alpha1/inference_graph.go:
```diff
@@ -0,0 +1,257 @@
+	NodeName string `json:"nodeName,omitempty"`
+
+	// named reference for InferenceService
+	ServiceName string `json:"serviceName,omitempty"`
```
I noticed that in the example InferenceGraph YAML files and doc snippets,  `service` is used instead of `serviceName` as declared here and in the generated CRD. Should YAML files and docs be updated with `serviceName`?

yuzisun replied:
updated

pvaneck commented on pkg/apis/serving/v1alpha1/inference_graph_validation.go:
```diff
@@ -0,0 +1,174 @@
+					i, route.StepName, nodeName, ig.Name)
+			}
+			if count != 1 {
+				return fmt.Errorf("Step %d (\"%s\") in node \"%s\" of InferenceGraph \"%s\" specifies more than one of nodeName, service, serviceUrl",
```
s/service/serviceName/ ?

pvaneck reviewed (commented):
Can you rebase on master to get the new github action e2e workflow?

yuzisun committed:
rename service to serviceName

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

njhill reviewed (approved):
Thanks @yuzisun !

pvaneck commented:
Thanks! Good to finally have this.
/lgtm

rachitchauhan43 commented on cmd/router/main.go:
```diff
@@ -37,32 +38,30 @@ import (
+	return body, err
 }
 
 func pickupRoute(routes []v1alpha1.InferenceStep) *v1alpha1.InferenceStep {
```
@yuzisun : Can you explain how this function is working ? How is it taking care of weighted distribution ?
